TFNetworkLayer.py

```diff
@@ -9411,3 +9411,344 @@ def get_layer_class_name_list():
   if not _LayerClassDictInitialized:
     _init_layer_class_dict()
   return sorted(_LayerClassDict.keys())
+
+
+class GatherPositionsLayer(LayerBase):
+  """
+  Gathers vectors at the specified positions on time axis.
+  """
+  layer_class = "gather_positions"
```

Comment:
This layer looks very much like the `GatherNdLayer`. What is the difference? This should be documented.
Also, if it is basically the same, then you should remove this, and just use `GatherNdLayer` instead. Or maybe it is easy to extend `GatherNdLayer` in such a way that you can use it.

Remember, the layers in Returnn should be simple, straight-forward, and not be specific for some specific task. See the Returnn documentation and introduction video lecture.

Reply:
I actually wasn't aware of the `GatherNdLayer` (since it was introduced long after I developed `GatherPositionsLayer`). Yes, they are quite similar, however, this layer allows for different indexing for each sequence by adding an offset to the indices before performing the gather operation on the flattened batch. And more importantly, it recovers the batch structure after the gather operation.
I'm not sure if this (especially the latter) can be done in a more generic way.
Also, I'm not sure if this functionality will be useful at all without `ProbabilisticMaskLayer`.

Reply:
But this sounds exactly like what `GatherNdLayer` does. It allows for different indexing for each sequence.